components/org.wso2.carbon.identity.oauth2.grant.organizationswitch/src/main/java/org/wso2/carbon/identity/oauth2/grant/organizationswitch/OrganizationSwitchGrant.java

```diff
@@ -99,10 +92,17 @@ public boolean validateGrant(OAuthTokenReqMessageContext tokReqMsgCtx) throws Id
         AuthenticatedUser authorizedUser = nonNull(tokenDO) ? tokenDO.getAuthzUser() :
                 AuthenticatedUser.createLocalAuthenticatedUserFromSubjectIdentifier(
                         validationResponseDTO.getAuthorizedUser());
+        if (StringUtils.equals(tokenRequestedTenantDomain, authorizedUser.getTenantDomain())) {
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("Provided token was already issued for the requested tenant domain: " +
+                        tokenRequestedTenantDomain);
+            }
+            throw new IdentityOAuth2Exception("Provided token was already issued for the requested tenant domain.");
+        }
```

Comment:
Can't we check the switching organization's relationship as well? whether the switching org and the initial token received org are in the same hierarchy

Reply:
Yes, better to have it. So we have more controlled switching.

Reply:
Fixed by - https://github.com/wso2-extensions/identity-oauth2-grant-organization-switch/pull/14/commits/a1887604fa5b79b1bc7702f8986a95f8f8aa9fc5

This commit includes some other improvements like 
- Using OSGI services instead of creating service objects.
- The method parameters can be changed https://github.com/wso2-extensions/identity-oauth2-grant-organization-switch/pull/14/files#diff-4582f045c1766b28a97c98c77e5dfd135e802b829bb8700e4d06c24a5bb2aef9R199